### DISCUSSION TITLE
representing a chain of computation steps

### DISCUSSION THREAD

**Discussion by mccraigmccraig** [koka-lang/koka#507] [Q&A]
I wanted a type to represent a chain of computation `step`s, where the output of one `step` forms the input to the next

each step has: `step<i,o,e>`, where `i` = input data, `o` = output data, `e` = any necessary effect

I came up with:

```
// it's a backwards cons of steps - last step first... 
// as steps are added to the chain, the interior step
// output/input types represented by j
// become invisible
pub type chain<i,o,e>
  First(first: step<i,o,e>)
  Link(last: some <j> step<j,o,e>, front: some <j> chain<i,j,e>)
```

which works, but it doesn't seem quite right - not least because the two `some <j>` statements introduce two separate type variables, but only one is needed

is there a better way of representing this ?

**Comment by woojamon** [comment]
Are you trying to write a monad? What are some examples of `step`s that you expect to have?

**Reply by mccraigmccraig** [reply]
well i wasn't thinking to write a monad... I'll use effects for the kind of things i would normally think of using a monad for... i was just playing around modeling a series of steps gathering data for a business computation, but it is a sequence of steps, so maybe that's a good way to be thinking of it
